### PR TITLE
Remove most reduces

### PR DESCRIPTION
**Pull request by chenglou**
I've split this into 10 diffs to show the changes more clearly. Feel free to browse them individually.

Essentially, there's just no good use-case for `reduce` (or `fold`), beside the trivial `reduce(+)` stuff in tutorials (which is really just `sum` and doesn't apply elsewhere).

Every `reduce` in this codebase can be written in a more understandable _and_ more performant way using stuff like `map`, `filter` and, splitting or recursion. There are a few left still, but those codepaths need to be completely changed for unrelated reasons.

Here are cases when `reduce` doesn't make sense (aka most of them):
- If you're just pushing/concatenating to the same accumulator array at the end anyway: just use `map`.
- If you're shrinking the array based on some conditions: use `filter` or `Belt.Array.keepMap`.
- If you're doing _any_ side-effect: if your `reduce` has side-effect then just use a for loop/`forEach`.
- If you're reducing to another collection: use recursion. Even recursion is more readable than `reduce`.

Performance-wise:
- The excessive allocations of `reduce` are killer. A few of them happen in hot loops, inside hotly reused components, inside hotly rerendered components. Btw I've seen React users immutably updating the accumulator collection in each step of `reduce`. What's the point...
- Notice a few `reduce`s got rewritten to `map(...)->join(char)`. This much more readable and actually better perf. `join` is specialized everywhere to allocate a predetermined large buffer and insert the pieces of string in the array in one shot (instead adding a few characters to a string in each `reduce` iteration and thrashing the string rope data structure).

Readability-wise:
- I'll let the result speak for itself.

Tldr try not to use reduce.

### CODE REVIEW

File: src/Playground.res [hunk 1]
```diff
@@ -1214,17 +1214,11 @@ module ControlPanel = {
           ("code", editorCode.current->LzString.compressToEncodedURIComponent),
         )->ignore
 
-        let querystring = Belt.Array.reduce(params, "", (acc, next) => {
-          let (key, value) = next
+        let querystring =
+          params->Js.Array2.map(((key, value)) => key ++ "=" ++ value)->Js.Array2.joinWith("&")
+          Js.log(querystring)
```
Comment (ryyppy) [inline src/Playground.res:1219]: Probably wanna remove this debugging `Js.Log`

File: src/components/Navigation.res [hunk 2]
```diff
@@ -493,26 +493,14 @@ let make = (~fixed=true, ~overlayState: (bool, (bool => bool) => unit)) => {
 
   let onStateChange = (~id, state) => {
     setCollapsibles(prev => {
-      Js.Array2.reduce(
```
Comment (ryyppy) [inline src/components/Navigation.res:496]: That's a good diff!

**Comment by ryyppy** [comment]
Thanks for the clear diffs! Didn't know about the crazy different perf characteristic between `reduce` vs `map` + `join` 👌

File: src/Playground.res [hunk 3]
```diff
@@ -1214,17 +1214,10 @@ module ControlPanel = {
           ("code", editorCode.current->LzString.compressToEncodedURIComponent),
         )->ignore
 
-        let querystring = Belt.Array.reduce(params, "", (acc, next) => {
-          let (key, value) = next
+        let querystring =
+          params->Js.Array2.map(((key, value)) => key ++ "=" ++ value)->Js.Array2.joinWith("&")
```
Comment (namenu) [inline src/Playground.res:1218]: Why `Js.Array2.map` instead of `Belt.Array.map`?
Belt is much [faster](https://johnridesa.bike/benchmark-bs/#list-array-map), obviously.

And... I see no chance of JS speeding up. (by looking at its polyfill)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#polyfill

Comment (vasco3) [reply src/Playground.res:1218]: good question

Comment (wontheone1) [reply src/Playground.res:1218]: > https://rescript-lang.org/docs/manual/latest/api/js

This might be the answer, but I am also tempted to vote for Belt.Array.map for the sake of the performance.

Comment (chenglou) [reply src/Playground.res:1218]: Either way's fine! The performance and readability gains of removing `reduce` here is enough that this doesn't matter. But yes, like wontheone1 linked, whenever possible we try to stay idiomatic and not carry a runtime. Here it's applied a bit inconsistently because some helpers don't exist in Js (e.g. `keepMap`).

If you're doing interop (like this repo is gonna do more of), Use the `Js.*` namespace. Otherwise `Belt` is indeed pretty good. The `map` is really just a for loop inside.
